extension/agenthealth/handler/stats/handler.go

```diff
@@ -0,0 +1,87 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package stats
+
+import (
+	"context"
+	"net/http"
+	"sync"
+
+	"github.com/amazon-contributing/opentelemetry-collector-contrib/extension/awsmiddleware"
+	"go.uber.org/zap"
+
+	"github.com/aws/amazon-cloudwatch-agent/extension/agenthealth/handler/stats/agent"
+	"github.com/aws/amazon-cloudwatch-agent/extension/agenthealth/handler/stats/client"
+	"github.com/aws/amazon-cloudwatch-agent/extension/agenthealth/handler/stats/provider"
+)
+
+const (
+	handlerID           = "cloudwatchagent.StatsHandler"
+	headerKeyAgentStats = "X-Amz-Agent-Stats"
+)
+
+func NewHandlers(logger *zap.Logger, cfg client.StatsConfig) ([]awsmiddleware.RequestHandler, []awsmiddleware.ResponseHandler) {
+	clientStats := client.NewHandler(cfg)
+	stats := newStatsHandler(logger, []agent.StatsProvider{clientStats, provider.GetProcessStats(), provider.GetFlagsStats()})
+	return []awsmiddleware.RequestHandler{clientStats, stats}, []awsmiddleware.ResponseHandler{clientStats, stats}
+}
+
+type statsHandler struct {
+	mu        sync.Mutex
+	logger    *zap.Logger
+	providers []agent.StatsProvider
+
+	headers map[string]string
+}
+
+func newStatsHandler(logger *zap.Logger, providers []agent.StatsProvider) *statsHandler {
+	sh := &statsHandler{
+		logger:    logger,
+		providers: providers,
+		headers:   make(map[string]string),
+	}
+	return sh
+}
+
+var _ awsmiddleware.RequestHandler = (*statsHandler)(nil)
+var _ awsmiddleware.ResponseHandler = (*statsHandler)(nil)
+
+func (sh *statsHandler) ID() string {
+	return handlerID
+}
+
+func (sh *statsHandler) Position() awsmiddleware.HandlerPosition {
+	return awsmiddleware.After
+}
+
+func (sh *statsHandler) HandleRequest(ctx context.Context, r *http.Request) {
+	header := sh.Header(awsmiddleware.GetOperationName(ctx))
+	if header != "" {
+		r.Header.Set(headerKeyAgentStats, header)
+	}
+}
+
+func (sh *statsHandler) HandleResponse(ctx context.Context, _ *http.Response) {
+	go sh.refreshHeader(awsmiddleware.GetOperationName(ctx))
```

Comment:
Do we call refresh on handling responses mainly for calculating latency?
Am I reading this write that we first need to have a response, which will populate the headers, that will then get set on the request handler (for subsequent request??) ?